JSON 문자열 분석기 6단계

popsmile opened the pull request:
- JSON 파일명을 argument로 받아 파싱 결과를 새로운 파일에 저장하는 기능을 `InputView` 와 `OutputView` 에 각각 추가했습니다. ( argument로 파일명만을 받아오기 위해 디렉토리 위치는 미리 상수로 지정해두었습니다. )
- JSON 파일 확장자를 `.gitignore` 에 추가해 읽고 새로 저장한 JSON 파일은 리모트 저장소에 올라가지 않도록 했습니다.

popsmile committed:
Add json extension to gitignore

popsmile committed:
Add functions to read arguments and contents of a file

popsmile committed:
Add a String extension to trim new lines

popsmile committed:
Add a function to write results as a file

popsmile committed:
Apply functions of reading and writing a file to main

godrm commented on JSONParser/JSONParser/InputView.swift:
```diff
@@ -10,7 +10,24 @@ import Foundation
 
 struct InputView {
     static private let guideMessage = "분석할 JSON 데이터를 입력하세요."
+    static private let inputFileDirPath = "/Users/yxxjy/DevNote/swift-jsonparser/"
```
코드에 이렇게 절대 경로를 지정하면 안됩니다. 
실행파일의 현재 경로를 구해서 상대 경로로 접근하세요.

godrm commented on JSONParser/JSONParser/InputView.swift:
```diff
@@ -10,7 +10,24 @@ import Foundation
+        guard let data = FileManager.default.contents(atPath: "\(inputFileDirPath)\(file)") else { return nil }
+        return String(decoding: data, as: UTF8.self)
+    }
+    
```
InputView가 CommandLine argument도 처리하고 파일도 읽고 readLine()도 다 하는것 같습니다. 분리하는게 어떨까요?
특히 CommandLine은 InputView가 의존하지 않도록 개선해보세요.

godrm commented on JSONParser/JSONParser/main.swift:
```diff
@@ -10,12 +10,16 @@ import Foundation
             return
         }
-        OutputView.showJSONPrettyPrinted(of: jsonDataForm)
+        OutputView.writeJSONPrettyPrinted(of: jsonDataForm)
```
요구사항에 파일이 있으면 파일에서 처리하고, 없으면 readInput()을 하도록 되어 있지 않던가요?
기존 기능을 그대로 유지하면서 확장하도록 개선해보세요.

popsmile committed:
Divide files to InputView, CommandLineReader and FileReader

Modify absolute path to relative path in FileReader. The relative path indicates user's Document directory

popsmile committed:
Fix the absolute path to a relative path of an output file

popsmile committed:
Create a protocol JSONParserExecutable

Make the struct Main to run whether extra arguments exist or not.

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -50,8 +48,8 @@ struct OutputView {
         let contents = jsonDataForm.prettyPrinted
-        let file = InputView.readArgument(atIndex: .outputFile) ?? defaultOutputFile
-        let fileURL = URL(fileURLWithPath: "\(outputFileDirPath)\(file)")
+        let file = CommandLineReader.readArgument(atIndex: .outputFile) ?? defaultOutputFile
```
OutputView에서 CommandLineReader 를 직접 접근하는 의존성을 갖는것과
외부에서 파일 이름을 넘겨주는 것과 어떤 차이가 있을까요? 
다른 객체에 의존하기 보다는 값을 넘겨주는 게 더 나은 선택일 수도 있습니다. 참고하세요.

popsmile replied:
네 ! 고민했던 부분, 정확히 짚어주셔서 감사합니다..ㅎㅎ 참고하겠습니다 :)

popsmile committed:
Move the constant of output file name to JSONFileParser

popsmile commented:
- 기존에 절대경로로 지정되었던 파일 경로를 상대경로로 수정했습니다. `FileManager` 의 `.documentDirectory` 를 사용했습니다.  
- `InputView` 에서 `argument` 를 다루던 부분과 `file` 을 다루던 부분을 각각 `CommandLineReader` 와 `FileReader` 로 분리했습니다. 
- 파일명 argument 의 유무에 상관없이 각각의 파싱 결과를 도출하도록 `protocol JSONParserExecutable` 을 만들어 `JSONLineParser` 와 `JSONFileParser` 가 이를 각각 따르도록 수정했습니다. 
- (마지막 피드백 반영사항) `OutputView` 내 `writeJSONPrettyPrinted()` 메소드를 호출하는 외부에서 파일이름을 넘기도록 수정했습니다.
